src/Workspaces/Remote/Core/RemoteHostAssetWriter.cs

```diff
@@ -0,0 +1,241 @@
+    {
+        // Create a channel to communicate between the searching and writing tasks.  This allows the searching task to
+        // find items, add them to the channel synchronously, and immediately continue searching for more items.
+        // Concurrently, the writing task can read from the channel and write the items to the pipe-writer.
+        var channel = Channel.CreateUnbounded<ChecksumAndAsset>(s_channelOptions);
+
+        // When cancellation happens, attempt to close the channel.  That will unblock the task writing the assets
+        // to the pipe. Capture-free version is only available on netcore unfortunately.
+        using var _ = cancellationToken.Register(
+#if NET
+            static (obj, cancellationToken) => ((Channel<ChecksumAndAsset>)obj!).Writer.TryComplete(new OperationCanceledException(cancellationToken)),
+            state: channel);
+#else
+            () => channel.Writer.TryComplete(new OperationCanceledException(cancellationToken)));
+#endif
+
+        // Spin up a task to go find all the requested checksums, adding results to the channel.
+        // Spin up a task to read from the channel, writing out the assets to the pipe-writer.
+        await Task.WhenAll(
+            FindAssetsFromScopeAndWriteToChannelAsync(channel.Writer, cancellationToken),
+            ReadAssetsFromChannelAndWriteToPipeAsync(channel.Reader, cancellationToken)).ConfigureAwait(false);
+    }
+
+    private async Task FindAssetsFromScopeAndWriteToChannelAsync(ChannelWriter<ChecksumAndAsset> channelWriter, CancellationToken cancellationToken)
+    {
+        Exception? exception = null;
+        try
+        {
+            await Task.Yield();
+
+            await _scope.FindAssetsAsync(
+                _assetPath, _checksums,
+                // It's ok to use TryWrite here.  TryWrite always succeeds unless the channel is completed. And the
+                // channel is only ever completed by us (after FindAssetsAsync completed) or if cancellation
+                // happens.  In that latter case, it's ok for writing to the channel to do nothing as we no longer
+                // need to write out those assets to the pipe.
+                static (checksum, asset, channelWriter) => channelWriter.TryWrite((checksum, asset)),
+                channelWriter, cancellationToken).ConfigureAwait(false);
+        }
+        catch (Exception ex) when ((exception = ex) == null)
+        {
+            throw ExceptionUtilities.Unreachable();
+        }
+        finally
+        {
+            // No matter what path we take (exceptional or non-exceptional), always complete the channel so the
+            // writing task knows it's done.
+            channelWriter.TryComplete(exception);
+        }
+    }
+
+    private async Task ReadAssetsFromChannelAndWriteToPipeAsync(ChannelReader<ChecksumAndAsset> channelReader, CancellationToken cancellationToken)
+    {
+        await Task.Yield();
```

Comment:
>await Task.Yield(); [](http://example.com/codeflow?start=8&length=19)

Trying to get my head around these Task.Yield calls. 

Out of curiosity, if the Task.WhenAll created the ReadAssetsFromChannelAndWriteToPipeAsync task first, would it make the Task.Yields unnecessary? (The ReadAssetsFromChannelAndWriteToPipeAsync task won't be able to hog the calling thread as it can't proceed past WaitToReadAsync unless the other method has hit a TryWrite or TryComplete call.

Reply:
the yielding is so that we don't do anything synchronously with these tasks on hte callsite.  i want to get the tasks immediately.

Reply:
i def think it could be argued this isn't necessary.  but i really like it for the clarity of: this is going to yield and go to the threadpool, and won't have any concerns about executing on the caller thread.

yes, we can infer that from some of the patterns, but a single yield makes it clear.

Reply:
Even if that would work, maybe it confuses things more that it's worth. There's enough asynchronicity in here that a couple Task.Yields probably isn't going to affect latency much.

Reply:
right.